tests/posix/rw/rw.cpp

```diff
@@ -667,6 +667,231 @@ TEST_F(rw, truncate)
+{
+	char buf[0x1000];
+	char buf00[sizeof(buf)];
+	char bufFF[sizeof(buf)];
+	PMEMfile *f;
+	ssize_t r;
+
+	memset(buf00, 0x00, sizeof(buf00));
+	memset(bufFF, 0xff, sizeof(bufFF));
+
+	f = pmemfile_open(pfp, "/file1", PMEMFILE_O_CREAT | PMEMFILE_O_RDWR,
+			  PMEMFILE_S_IRWXU);
+	ASSERT_NE(f, nullptr) << strerror(errno);
+
+	EXPECT_TRUE(test_pmemfile_stats_match(pfp, 2, 0, 0, 0, 0));
+
+	/* Allocate a range, file size is expected to remain zero */
+	r = pmemfile_fallocate(pfp, f, PMEMFILE_FL_KEEP_SIZE, 0x1000, 0x10000);
+	ASSERT_EQ(r, 0) << COND_ERROR(r);
+	ASSERT_EQ(test_pmemfile_path_size(pfp, "/file1"), 0);
+
+	/*
+	 * Allocated a 64K range, expecting a large block, or 16 pieces
+	 * of 4K blocks
+	 */
+	EXPECT_TRUE(test_pmemfile_stats_match(
+		pfp, 2, 0, 0, 0, (env_block_size == 4096) ? 16 : 1));
+
+	/*
+	 * Allocate the same range, file size is expected to change,
+	 * but no new allocation should happen.
+	 */
+	r = pmemfile_fallocate(pfp, f, 0, 0x1000, 0x10000);
+	ASSERT_EQ(r, 0) << COND_ERROR(r);
+	ASSERT_EQ(test_pmemfile_path_size(pfp, "/file1"), 0x1000 + 0x10000);
+
+	EXPECT_TRUE(test_pmemfile_stats_match(
+		pfp, 2, 0, 0, 0, (env_block_size == 4096) ? 16 : 1));
+
+	/*
+	 * Now remove an interval, that overlaps with the previously
+	 * allocated interval.
+	 * This should be rounded to the interval: [0 - 0x4000[ - thus
+	 * removing 3 pieces of 4K blocks, or just zeroing out some data.
+	 */
+
+	/* But first make sure it is not allowed without the KEEP_SIZE flag */
+	r = pmemfile_fallocate(pfp, f, PMEMFILE_FL_PUNCH_HOLE, 0x0007, 0x4123);
+	ASSERT_EQ(r, -1);
+	ASSERT_EQ(errno, EINVAL);
+
+	r = pmemfile_fallocate(pfp, f,
+			       PMEMFILE_FL_PUNCH_HOLE | PMEMFILE_FL_KEEP_SIZE,
+			       0x0007, 0x4123);
+	ASSERT_EQ(r, 0) << COND_ERROR(r);
+	ASSERT_EQ(test_pmemfile_path_size(pfp, "/file1"), 0x1000 + 0x10000);
+
+	EXPECT_TRUE(test_pmemfile_stats_match(
+		pfp, 2, 0, 0, 0, (env_block_size == 4096) ? 13 : 1));
+
+	/*
+	 * Writing some bytes -- this should allocate two new blocks when
+	 * the block size is fixed to 4K bytes.
+	 */
+	static constexpr char data0[] = "testing testy tested tests";
+	static constexpr ssize_t l0 = sizeof(data0) - 1;
+
+	ASSERT_EQ(pmemfile_lseek(pfp, f, 0x1ffe, PMEMFILE_SEEK_SET), 0x1ffe);
+	ASSERT_EQ(pmemfile_write(pfp, f, data0, l0), l0);
+
+	EXPECT_TRUE(test_pmemfile_stats_match(
+		pfp, 2, 0, 0, 0, (env_block_size == 4096) ? 13 + 2 : 1));
+
+	/*
+	 * Try to read the test data, there should be zeroes around it.
+	 */
+	ASSERT_EQ(pmemfile_lseek(pfp, f, 0x1ffd, PMEMFILE_SEEK_SET), 0x1ffd);
+	memset(buf, 0xff, sizeof(buf));
+	r = pmemfile_read(pfp, f, buf, sizeof(buf));
+	ASSERT_EQ(r, (ssize_t)sizeof(buf)) << COND_ERROR(r);
+	ASSERT_EQ(buf[0], '\0');
+	ASSERT_EQ(memcmp(buf + 1, data0, l0), 0);
+	ASSERT_EQ(memcmp(buf + 1 + l0, buf00, sizeof(buf) - 1 - l0), 0);
+
+	/*
+	 * Punch a hole at [0x1fff, 0x4777[ interval, which should be
+	 * internally translated to the [0x2000, 0x4000[ interval.
```

Comment:
Or what about ``[0x1fff, 0x4777)`` ?
I remember using this at school, perhaps it was ``<0x1fff, 0x4777)`` in Polish schools?
I also remember using ``j`` for the imaginary unit instead of ``i``, so yes, some teachers have their own ideas...